src/libmeasurement_kit/engine.cpp

```diff
@@ -0,0 +1,480 @@
+        active_ = true;
+    }
+
+    void signal() {
+        {
+            std::unique_lock<std::mutex> _{mutex_};
+            active_ = false;
+        }
+        cond_.notify_one();
+    }
+
+    ~Semaphore() = default;
+
+  private:
+    bool active_ = false;
+    std::condition_variable cond_;
+    std::mutex mutex_;
+};
+
+class TaskImpl {
+  public:
+    std::condition_variable cond;
+    std::deque<nlohmann::json> deque;
+    std::atomic_bool interrupted{false};
+    std::mutex mutex;
+    SharedPtr<Reactor> reactor = Reactor::make();
+    std::atomic_bool running{false};
+    std::thread thread;
+};
+
+static void task_run(TaskImpl *pimpl, nlohmann::json &settings);
+static bool is_event_valid(const std::string &);
+
+static void emit(TaskImpl *pimpl, nlohmann::json &&event) {
+
+    // In debug mode, make sure that we're emitting an event that we know
+    assert(event.is_object());
+    assert(event.count("type") != 0);
+    assert(event.at("type").is_string());
+    assert(is_event_valid(event.at("type").get<std::string>()));
+
+    // Actually emit the event.
+    {
+        std::unique_lock<std::mutex> _{pimpl->mutex};
+        pimpl->deque.push_back(std::move(event));
+    }
+    pimpl->cond.notify_all(); // more efficient if unlocked
+}
+
+Task::Task(nlohmann::json &&settings) {
+    pimpl_ = std::make_unique<TaskImpl>();
+    std::promise<void> barrier;
+    std::future<void> started = barrier.get_future();
+    pimpl_->thread = std::thread([this, &barrier,
+                                         settings = std::move(
+                                                 settings)]() mutable {
+        pimpl_->running = true;
+        barrier.set_value();
+        static Semaphore semaphore;
+        semaphore.wait(); // block until a previous task has finished running
+        task_run(pimpl_.get(), settings);
+        pimpl_->running = false;
+        pimpl_->cond.notify_all(); // tell the readers we're done
+        semaphore.signal();        // allow another task to run
+    });
+    started.wait(); // guarantee Task() completes when the thread is running
+}
+
+bool Task::is_running() const {
+    return pimpl_->running; // is an atomic var
+}
+
+void Task::interrupt() {
+    // both variables are safe to use in a MT context
+    pimpl_->reactor->stop();
+    pimpl_->interrupted = true;
+}
+
+nlohmann::json Task::wait_for_next_event() {
+    std::unique_lock<std::mutex> lock{pimpl_->mutex};
+    // purpose: block here until we stop running or we have events to read
+    pimpl_->cond.wait(lock, [this]() { //
+        return !pimpl_->running || !pimpl_->deque.empty();
+    });
+    // must be first so we drain the queue before emitting the final null
+    if (!pimpl_->deque.empty()) {
+        auto rv = std::move(pimpl_->deque.front());
+        pimpl_->deque.pop_front();
+        return nlohmann::json{std::move(rv)};
+    }
+    assert(!pimpl_->running);
+    return nlohmann::json(); // this is a `null` JSON object
```

Comment:
A null json object is a null object or a json object with empty field?

Reply:
The former thing!

Reply:
Based on @willscott's suggestion, now one is not required to check whether the returned JSON is null, but rather he/she can use `is_running()` to check whether things are working. Yet, I believe a `null` JSON string is returned in any case when the queue hits the end. I am not sure how I can avoid this, tho, since I need anyway to return something to the caller if I'm blocked in reading the queue.